src/axolotl/utils/callbacks.py

```diff
@@ -317,3 +321,177 @@ def on_evaluate(
+                        else:
+                            pos_ranges = find_ranges(pos_ids)
+
+                        for pos_range in pos_ranges:
+                            start, end = pos_range
+                            if start == end:
+                                continue
+
+                            input_ids = input_ids_all[start:end+1]
+                            labels = labels_all[start:end+1]
+                            # input_ids[start:end] = tokenizer.pad_token_id
+
+                            tokens_without_loss = (labels == IGNORE_INDEX)
+                            tokens_with_loss = (labels != IGNORE_INDEX)
+                            tokens_exclude_padding = (input_ids != tokenizer.pad_token_id)
+
+                            prompt_token_includes = tokens_without_loss & tokens_exclude_padding
+
+                            prompt_token_ids = input_ids[prompt_token_includes]
+                            prompt_token_ids_list.append(prompt_token_ids)
+
+                            completion_token_ids = input_ids[tokens_with_loss]
+                            completion_token_ids_list.append(completion_token_ids)
+                            # completion_text = tokenizer.decode(completion_token_ids)
+                            # completion_texts.append(completion_text)
+
+                            # completion_logit = logits[tokens_with_loss]
+                            # predicted_tokens = logits_to_tokens(completion_logit)
+                            # prediction_text = tokenizer.decode(predicted_tokens)
+                            # prediction_texts.append(prediction_text)
+
+                    prompt_texts = tokenizer.batch_decode(prompt_token_ids_list, skip_special_tokens=True)
+                    completion_texts = tokenizer.batch_decode(completion_token_ids_list, skip_special_tokens=True)
+
+                    with torch.no_grad():
+                        generation_config = GenerationConfig(
+                            repetition_penalty=1.1,
+                            max_new_tokens=32,
+                            bos_token_id=tokenizer.bos_token_id,
+                            eos_token_id=tokenizer.eos_token_id,
+                            pad_token_id=tokenizer.pad_token_id,
+                            do_sample=False,
+                            use_cache=True,
+                            return_dict_in_generate=True,
+                            output_attentions=False,
+                            output_hidden_states=False,
+                            output_scores=False,
+                        )
+
+                        encoding = tokenizer(prompt_texts, padding=True, return_tensors='pt').to(self.cfg.device)
+                        new_predictions = trainer.model.generate(**encoding, generation_config=generation_config) # FIXME: when sample_packing=True then error: "TypeError: varlen_fwd(): incompatible function arguments."
```

Comment:
**Help Wanted**

✅ I think `sample_packing: false` is working, however, need more testing. If you could test that'd be great!
❌ `sample_packing: true` works up until the `training.model.generate()` call, then error . Something about packing + flash attention, it no longer behaves the same / expects the same args.


```
varlen_fwd(): incompatible function arguments. The following argument types are supported:
    1. (arg0: torch.Tensor, arg1: torch.Tensor, arg2: torch.Tensor, arg3: Optional[torch.Tensor], arg4: torch.Tensor, arg5: torch.Tensor, arg6: int, arg7: int, arg8: float, arg9: float, arg10: bool, arg11: bool, arg12: bool, arg13: Optional[torch.Generator]) -> List[torch.Tensor]

Invoked with: tensor([[[-0.0371, -0.0184, -0.0273,  ...,  0.1377,  0.2227,  0.1807],

...

        [  0, 348],
        [  0, 348]], device='cuda:0', dtype=torch.int32), tensor([348, 348, 348, 348, 348, 348, 348, 348, 348, 348, 348, 348, 348, 348,
        348, 348, 348, 348, 348, 348, 348, 348, 348, 348, 348, 348, 348, 348,
        348, 348, 348, 348, 348, 348, 348, 348, 348, 348, 348, 348, 348, 348,
        348, 348, 348, 348, 348, 348, 348, 348, 348, 348, 348, 348, 348],
       device='cuda:0'), tensor([348, 348, 348, 348, 348, 348, 348, 348, 348, 348, 348, 348, 348, 348,
        348, 348, 348, 348, 348, 348, 348, 348, 348, 348, 348, 348, 348, 348,
        348, 348, 348, 348, 348, 348, 348, 348, 348, 348, 348, 348, 348, 348,
        348, 348, 348, 348, 348, 348, 348, 348, 348, 348, 348, 348, 348],
       device='cuda:0'), 0.0, 0.125, False, True, False, None
  File "/home/user/axolotl/src/axolotl/monkeypatch/llama_attn_hijack_flash.py", line 205, in flashattn_forward
    output = flash_attn_varlen_qkvpacked_func(
  File "/home/user/axolotl/src/axolotl/monkeypatch/llama_attn_hijack_flash.py", line 606, in forward
    hidden_states, self_attn_weights, present_key_value = self.self_attn(
  File "/home/user/axolotl/src/axolotl/monkeypatch/llama_attn_hijack_flash.py", line 528, in llama_model_forward
    layer_outputs = decoder_layer(
  File "/home/user/axolotl/src/axolotl/utils/callbacks.py", line 449, in log_table_from_dataloader
    new_predictions = trainer.model.generate(**encoding, generation_config=generation_config)
  File "/home/user/axolotl/src/axolotl/utils/callbacks.py", line 476, in on_evaluate
    log_table_from_dataloader("Eval", eval_dataloader)
  File "/home/user/axolotl/src/axolotl/train.py", line 118, in train
    trainer.train(resume_from_checkpoint=resume_from_checkpoint)
  File "/home/user/axolotl/scripts/finetune.py", line 279, in do_cli
    train(cfg=parsed_cfg, cli_args=parsed_cli_args, dataset_meta=dataset_meta)
  File "/home/user/axolotl/scripts/finetune.py", line 283, in <module> (Current frame)
    fire.Fire(do_cli)
TypeError: varlen_fwd(): incompatible function arguments. The following argument types are supported:
    1. (arg0: torch.Tensor, arg1: torch.Tensor, arg2: torch.Tensor, arg3: Optional[torch.Tensor], arg4: torch.Tensor, arg5: torch.Tensor, arg6: int, arg7: int, arg8: float, arg9: float, arg10: bool, arg11: bool, arg12: bool, arg13: Optional[torch.Generator]) -> List[torch.Tensor]

Invoked with: tensor([[[-0.0371, -0.0184, -0.0273,  ...,  0.1377,  0.2227,  0.1807],
         [ 0.0194, -0.0742, -0.0518,  ...,  0.2129,  0.1406,  0.1934],

```